src/prpl.ts

```diff
@@ -12,15 +12,175 @@
  * http://polymer.github.io/PATENTS.txt
  */
 
+import * as fs from 'fs';
 import * as http from 'http';
+import * as path from 'path';
 import * as send from 'send';
 import * as url from 'url';
 
-export function handler(root?: string): (
+import * as capabilities from './capabilities';
+import * as push from './push';
+
+// The subset of the polymer.json specification that we care about for serving.
+// https://www.polymer-project.org/2.0/docs/tools/polymer-json
+// https://github.com/Polymer/polymer-project-config/blob/master/src/index.ts
+export interface ProjectConfig {
+  entrypoint?: string;
+  builds?: {
+    name?: string,
+    browserCapabilities?: capabilities.BrowserCapability[],
+  }[];
+}
+
+// Matches "/foo/bar.png" but not "/foo.png/bar".
+const hasFileExtension = new RegExp('\\.[^/]*$');
+
+// TODO Service worker location should be configurable.
+const isServiceWorker = new RegExp('service-worker.js$');
+
+/**
+ * Return a new HTTP handler to serve a PRPL-style application.
+ */
+export function handler(rootDir?: string, config?: ProjectConfig): (
     request: http.IncomingMessage, response: http.ServerResponse) => void {
-  root = root || '.';
+  const root = rootDir || '.';
+  console.info(`serving files from "${root}"`);
+  const builds = loadBuilds(config, root);
+
   return function(request, response) {
-    const pathname = url.parse(request.url || '').pathname || '/';
-    send(request, pathname, {root: root}).pipe(response);
+    // Serve the entrypoint for the root path, and for all other paths that
+    // don't have a corresponding static resource on disk. As a special
+    // case, paths with file extensions are always excluded because they are
+    // likely to be not-found static resources rather than application
+    // routes.
+    const pathname = url.parse(request.url || '/').pathname || '/';
+    const serveEntrypoint = pathname === '/' ||
+        (!hasFileExtension.test(pathname) &&
+         !fs.existsSync(path.join(root, pathname)));
+
+    // Find the highest ranked build suitable for this user agent.
+    const clientCapabilities =
+        capabilities.browserCapabilities(request.headers['user-agent']);
+    const build = builds.find((b) => b.canServe(clientCapabilities));
+
+    // We warned about this at startup. You should probably provide a fallback
+    // build with no capabilities, at least to nicely inform the user. Note
+    // that we only return this error for the entrypoint; we always serve fully
+    // qualified static resources.
+    if (!build && serveEntrypoint) {
+      response.writeHead(400);
+      response.end('This browser is not supported.');
+      return;
+    }
+
+    const sendFile = (build && serveEntrypoint) ? build.entrypoint : pathname;
```

Comment:
fileToSend is clearer to me. sendFile makes me think of http://man7.org/linux/man-pages/man2/sendfile.2.html